### DISCUSSION TITLE
How to use newrelic agent in nextjs app

### DISCUSSION THREAD

**Discussion by nitintecg** [awslabs/aws-lambda-web-adapter#401] [Q&A]
Hello Team,
I want to know how to use newrelic agent in nextjs app in lambda using this layer. Can you provide the guide here as i have tried many ways but newrelic agent is not sending data to newrelic

**Comment by bnusunny** [comment]
Could you share the doc link for the newrelic agent you tried?

**Comment by nitintecg** [comment]
I have followed https://newrelic.com/blog/how-to-relic/nextjs-monitor-application-data documentation. I have modified run.sh file to 
`node -r @newrelic/next server.js` since we are using nextjs app standalone build.

**Comment by bnusunny** [comment]
I don't have New Relic account to experiment. Could you try New Relic Lambda Extension? 

https://docs.newrelic.com/docs/infrastructure/amazon-integrations/aws-integrations-list/aws-lambda-telemetry/

**Comment by nitintecg** [comment]
sure i will try this 
thanks

**Comment by nitintecg** [comment]
@bnusunny i have tried this approach this is not working

**Comment by bnusunny** [comment]
Did you see any error messages?

**Comment by nitintecg** [comment]
im not getting error messages and its giving response but in newrelic platform data is not there

**Comment by jeffknaide-vt** [comment]
@bnusunny we are trying to do something similar with a python+fastapi streaming application. we are using the NewRelic lambda layer along with the lambda web adapter layer. When using the bootstrap function as directed in the lambda web adapter docs, we encounter the following error:
```
/opt/bootstrap: line 3: /var/task/newrelic_lambda_wrapper.handler: No such file or directory
```
The NR lambda layer wraps the user defined handler and expects to be executed as a python file. I have experimented with providing a custom bootstrap file to trigger the NR lambda wrapper function as a python function. something like 
```
exec -- python -m ${_HANDLER}
```
but this leads to a different error
```
/var/lang/bin/python: Error while finding module specification for 'newrelic_lambda_wrapper.handler' (ModuleNotFoundError: No module named 'newrelic_lambda_wrapper')
```
i've unsuccessfully experimented with a handful of different formulations of how to call the handler, along with trying to find the handler function in the lambda runtime environment to explicitly call it out, updating the pythonpath/path, etc.

The other potential issue here is that the NR lambda wrapper is going to expect a python function as the user defined handler, not a bash script. I suppose we may be able to get around this by starting up our uvicorn server in a python file?

do you have any familiarity with how the lambda web adapter layer interacts with other layers that wrap code at runtime, and the best path to connect them?

**Comment by bnusunny** [comment]
This might be a wrong direction. With LWA, the web app is a stand-alone process, like a normal web application. We should try to use the normal instrument approach for web apps.

**Comment by nitintecg** [comment]
I have tried below solution and it worked in newrelic without installing any new layer
https://github.com/newrelic/newrelic-lambda-layers?tab=readme-ov-file#manual-instrumentation-for-es-modules

**Comment by jeffknaide-vt** [comment]
@bnusunny got it - thanks! I've tried using the NR approach for a FastAPI server (described [here](https://docs.newrelic.com/install/python/?python-non-web=non-web-no&python-docker=docker-no&python-framework=fastapi&python-server=uvicorn)), without success. The NR docs suggest starting up the server with the NR admin utility as 
```
newrelic-admin run-program uvicorn path_to_app
```
I've adapted this in our run.sh file, with some of the formatting suggested in the web-adapter docs 
```
PATH=$PATH:$LAMBDA_TASK_ROOT/bin PYTHONPATH=$LAMBDA_TASK_ROOT exec newrelic-admin run-program uvicorn path_to_app
```

this results in a few strange logs 
```
/var/task/path/to/run.sh: /var/task/bin/newrelic-admin: /opt/hostedtoolcache/Python/3.9.12/x64/bin/python3.9: bad interpreter: No such file or directory
/opt/bootstrap: line 3: /var/task/path/to/run.sh: Success
```
Not sure why we're seeing both error and success logs there. Does the file path from the log (`/opt/hostedtoolcache/Python/3.9.12/x64/bin/python3.9`) appear to point at the right python interpreter for the web adapter? If so, perhaps the issue is around how we are installing the NR agent dependency.

**Comment by bnusunny** [comment]
`/opt/hostedtoolcache/Python/3.9.12/x64/bin/python3.9` - this is not where python interpreter is installed in Python runtime. To use the python interpreter in the managed runtime, just use `python` since it is on the PATH.
